### ISSUE TITLE
`shell` command makes tcp pivot implants unreachable

### ISSUE THREAD

**Issue by spencerdodd** [BishopFox/sliver#411]
**Describe the bug**

attempting to run a `shell` command on a sliver that utilizes a tcp pivot listener as its upstream callback will result in making the implant completely unreachable. before issuing the `shell` command, the implant will behave fine, but afterwards, it will timeout all requests and become unresponsive. 

**To Reproduce**
1. Generate and execute a sliver with an `http` callback to the server
2. spin up a tcp pivot listener on that sliver
3. generate and execute a new sliver that targets the pivot listener as it's callback
4. interacting with the second, nested, sliver: run `ps` or another command and see good output
5. interacting with the second, nested, sliver: run `shell`
6. see nested sliver become unresponsive

**Expected behavior**

running the `shell` command on a nested sliver implant should not result in the implant becoming unresponsive. as long as you avoid the `shell` command, a nested implant behaves as expected. I've seen this behavior with both linux and windows implants as both the pivot listener and the nested implant.

**Desktop (please complete the following information):**

```
sliver (ICY_GYMNASTICS) > version

[*] Client v1.4.14 - 6c02971b54831884d30407b632a379947dd289ad - linux/amd64
    Compiled at 2021-04-22 19:04:25 -0400 EDT
    Compiled with go version go1.16.3 linux/amd64


[*] Server v1.4.14 - 6c02971b54831884d30407b632a379947dd289ad - linux/amd64
    Compiled at 2021-04-22 19:04:25 -0400 EDT
```

**Additional context**

```
sliver (ICY_GYMNASTICS) > ps
...[snipped]...
8024  3796  ICY_GYMNASTICS.exe           DESKTOP-IKFU76H\user  1        

sliver (ICY_GYMNASTICS) > shell

? This action is bad OPSEC, are you an adult? Yes
[*] Opening shell tunnel (EOF to exit) ...

[!] rpc error: code = Unknown desc = Implant timeout
Shell exited

sliver (ICY_GYMNASTICS) > ps

[!] rpc error: code = Unknown desc = Implant timeout
```

and I believe the relevant `~/.sliver/logs/sliver.log` entries

```
[36mINFO[0m[2021-04-27T20:19:44-04:00] [sliver/server/c2/tcp-http.go:309] 192.168.152.106:45928 - /static/underscore.min.js?_=5341073 - Mozilla/5.0 (Windows NT 10.0; WOW64; Trident/7.0; rv:11.0) like Gecko 
[36mINFO[0m[2021-04-27T20:19:44-04:00] [sliver/server/c2/tcp-http.go:309] 192.168.152.106:45926 - /signin.php?_=84579065 - Mozilla/5.0 (Windows NT 10.0; WOW64; Trident/7.0; rv:11.0) like Gecko 
[36mINFO[0m[2021-04-27T20:19:44-04:00] [github.com/grpc-ecosystem/go-grpc-middleware@v1.2.2/logging/logrus/options.go:211] finished unary call with code OK 
[36mINFO[0m[2021-04-27T20:19:46-04:00] [github.com/grpc-ecosystem/go-grpc-middleware@v1.2.2/logging/logrus/options.go:211] finished unary call with code OK 
[36mINFO[0m[2021-04-27T20:19:48-04:00] [github.com/grpc-ecosystem/go-grpc-middleware@v1.2.2/logging/logrus/options.go:211] finished unary call with code OK 
[36mINFO[0m[2021-04-27T20:19:49-04:00] [sliver/server/c2/tcp-http.go:309] 192.168.152.106:45928 - /static/underscore.min.js?_=84597934 - Mozilla/5.0 (Windows NT 10.0; WOW64; Trident/7.0; rv:11.0) like Gecko 
[31mERRO[0m[2021-04-27T20:20:56-04:00] [github.com/grpc-ecosystem/go-grpc-middleware@v1.2.2/logging/logrus/options.go:215] finished unary call with code Unknown 
[37mDEBU[0m[2021-04-27T20:20:56-04:00] [sliver/server/c2/tcp-http.go:512] Poll time out 
```

Please let me know if you need more info or cannot reproduce. Thanks!

**Comment by spencerdodd** [comment]
some more info: it does appear to spin up the sacrificial shell on the host (linux example)

```
sliver (ENDLESS_EARRINGS) > shell

? This action is bad OPSEC, are you an adult? Yes
[*] Opening shell tunnel (EOF to exit) ...

[*] Started remote shell with pid 13321
```

```
$ ps aux 
...[snipped]...
root      13321  0.0  0.1   6992  3640 pts/3    Ss+  20:51   0:00 /bin/bash
```

**Comment by spencerdodd** [comment]
I have had instances where it locks up the entire client instance when the `shell` command hangs and the instance must be restarted. I am running these commands direct on the server cli interface without utilizing a separate client user/binary.

**Comment by rkervella** [comment]
I ran a quick test and it seems the `shell` commands works sporadically in general (not just for pivots), sometimes it just blocks the implant, sometimes it simply works. I suspect something's acting weird in the tunnel code.

**Comment by rkervella** [comment]
Alright, looks like the implants receive the events properly, but somehow don't respond to the server. Must be some kind of read/write lock in the pivot code.

**Comment by moloch--** [comment]
Fixed in v1.5
